xarray/core/dataarray.py

```diff
@@ -158,6 +158,28 @@ def _infer_coords_and_dims(
     return new_coords, dims
 
 
+def _check_data_shape(data, coords, dims):
+    if data is dtypes.NA:
+        data = np.nan
+    if (
+        coords is not None
+        and utils.is_scalar(data, include_0d=False)
+        and not isinstance(data, indexing.ExplicitlyIndexed)
```

Comment:
I'm a little surprise that this check on this line is necessary -- what sort of errors do you get if you don't include it? I would expect that everything that is an instance of `ExplicitlyIndexed` would already fail the `utils.is_scalar(data, include_0d=False)` check.

Reply:
I found that `xr.open_rasterio` would fail in `test_backends.py`. Not sure if this is the best way to debug but I sneaked some print statements to verify what kind of object was getting past `is_scalar` and found it was `<class 'xarray.core.indexing.MemoryCachedArray'>`.
After more digging and realizing `MemoryCachedArray` inherits from `ExplictlyIndexed`, I wrote a simple test for it on `test_indexing.py` inside `TestMemoryCachedArray`

```python
def test_is_scalar(self):
    original = np.full((4, 3, 3), 1)
    wrapped = indexing.MemoryCachedArray(original)
    assert not utils.is_scalar(wrapped, include_0d=False)
```
And it failed. The reason might be because `is_scalar` only checks if an object has `__iter__`, `__array_function__` or if its a dask array and `MemoryCachedArray` does not meet that criteria, hence is considered a scalar
https://github.com/pydata/xarray/blob/131f6022763d35edf461b11857d7c4ec6630b19d/xarray/core/utils.py#L279-L291

The solution I came up with was from how `as_compatible_data` checks for this type of object
https://github.com/pydata/xarray/blob/131f6022763d35edf461b11857d7c4ec6630b19d/xarray/core/variable.py#L37-L40

Reply:
Thanks for the clear description @DangoMelon and for working through all that! 

I think for now we should add a check for `MemoryCachedArray` into `is_scalar` and remove the check here. 

Any thoughts @shoyer ?

Reply:
Yes, thanks for tracking this down! I agree with Max, let’s put this check (for ExplicitlyIndexed) into is_scalar instead.